crates/exex/exex/src/dyexex.rs

```diff
@@ -0,0 +1,188 @@
+//! Type-safe abstractions for Dynamically Loaded ExExes
+
+use std::{
+    env::consts::{DLL_PREFIX, DLL_SUFFIX},
+    fmt::Debug,
+    future::Future,
+    path::Path,
+    pin::Pin,
+    sync::Arc,
+};
+
+use eyre::{OptionExt, Result};
+use futures::future::BoxFuture;
+use libloading::{Library, Symbol};
+
+use crate::ExExContextDyn;
+
+/// ExEx launch function dynamic library symbol name.
+const LAUNCH_EXEX_FN: &[u8] = b"_launch_exex";
+
+/// Regarding ExEx launch entrypoint future
+type ExExFut = BoxFuture<'static, BoxFuture<'static, Result<()>>>;
+
+/// Dynamically loads an ExEx entrypoint, which accepts a user-defined function representing the
+/// core ExEx logic. The provided function must take an [`ExExContextDyn`](`crate::ExExContextDyn`)
+/// as its argument.
+///
+/// # Returns
+/// A Future that will be polled by the [`ExExManager`](`crate::ExExManager`).
+///
+/// ## Example usage:
+/// ```rust
+/// use futures::TryStreamExt;
+///
+/// use reth_exex::{define_exex, ExExContextDyn, ExExEvent, ExExNotification};
+/// use reth_tracing::tracing::info;
+///
+/// /// An ExEx is just a future, which means you can implement all of it in an async function!
+/// ///
+/// /// This ExEx just prints out whenever either a new chain of blocks being added, or a chain of
+/// /// blocks being re-orged. After processing the chain, emits an [ExExEvent::FinishedHeight] event.
+/// async fn exex(mut ctx: ExExContextDyn) -> eyre::Result<()> {
+///     while let Some(notification) = ctx.notifications.try_next().await? {
+///         match &notification {
+///             ExExNotification::ChainCommitted { new } => {
+///                 info!(committed_chain = ?new.range(), "Received commit");
+///             }
+///             ExExNotification::ChainReorged { old, new } => {
+///                 info!(from_chain = ?old.range(), to_chain = ?new.range(), "Received reorg");
+///             }
+///             ExExNotification::ChainReverted { old } => {
+///                 info!(reverted_chain = ?old.range(), "Received revert");
+///             }
+///         };
+///
+///         if let Some(committed_chain) = notification.committed_chain() {
+///             ctx.events.send(ExExEvent::FinishedHeight(committed_chain.tip().num_hash()))?;
+///         }
+///     }
+///
+///     Ok(())
+/// }
+///
+/// // Use the macro to generate the entrypoint function
+/// define_exex!(exex_init);
+/// ```
+#[macro_export]
+macro_rules! define_exex {
+    ($user_fn:ident) => {
+        #[no_mangle]
+        pub extern "Rust" fn _launch_exex(
+            ctx: $crate::ExExContextDyn,
+        ) -> futures::future::BoxFuture<
+            'static,
+            futures::future::BoxFuture<'static, eyre::Result<()>>,
+        > {
+            futures::FutureExt::boxed(async move {
+                Box::pin(async move { exex(ctx).await })
+                    as futures::future::BoxFuture<'static, eyre::Result<()>>
+            })
```

Comment:
not sure i understand what's going on here, it should be more like https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.map